File: apps/billing/lib/db/queries/history.ts
// IMPORTED HISTORY: the only writer of `billing.history`, and its reads.
//
// ===========================================================================
// THE AGENT MAPS, THIS APP STORES AND SHOWS
// ===========================================================================
// There is no importer UI and no mapping engine — BRIEF.md forbids both by
// name. An agent reads a Zoho or Invoicely export, maps each bill onto
// `ImportHistoryRow`, and posts the rows. This file checks them and writes them.
// It never guesses: a field it cannot accept is a refusal naming the row and
// the field, and an ambiguity the MAPPER could not resolve arrives as an import
// flag, in words, and is shown forever.
//
// ===========================================================================
// ALL OR NOTHING, AND A DUPLICATE IS A 409, NOT A SKIP
// ===========================================================================
// One import is one transaction. If any row is refused, nothing is written, and
// the refusal lists every problem it found — not the first — because a mapper
// fixing a 400-row file one error per round trip is a mapper that gives up.
//
// A row whose `(source, source_ref)` is already in the archive refuses the whole
// import with a 409 naming the existing row's #number. Not an upsert and not a
// merge: two exports disagreeing about one bill is something a person has to
// look at, and silently skipping it would make "imported 14" mean "imported 11
// and quietly ignored 3". Re-running the same file therefore changes nothing and
// says exactly which rows were already there.
//
// The unique index `uq_history_source_ref` is the MECHANISM; the read before the
// insert only turns what would be a bare constraint error into a sentence. Two
// imports racing past that read both reach the insert, one commits, and the
// other's unique violation is caught here and answered with the same 409 — so
// twenty concurrent imports of one row insert exactly one.
//
// ===========================================================================
// `source_ref` IS NEVER TOUCHED
// ===========================================================================
// Not trimmed, not case-folded, not reformatted. It is the only handle back to
// the system of record. Every other text field is checked for blankness without
// being rewritten either; nothing on an imported row is normalised, because the
// row's value is that it says what the source said.

import { and, desc, eq, inArray, sql } from 'drizzle-orm'
import { billingCompany, billingHistory, users } from '../schema'
import { getDb } from '../client'
import { allocateSeqBlock } from './seq'
import { todayInZurich } from './lifecycle'
import type { WriteCtx } from './invoices'
import { HISTORY_SOURCES, HISTORY_STATUSES } from '@/lib/vocabularies'
import { HISTORY_LIMITS, LIST_LIMIT_DEFAULT, LIST_LIMIT_MAX } from '@/lib/limits'
import type {
  ActorVia,
  HistoryEntry,
  HistorySource,
  HistoryStatus,
  ImportHistoryResult,
  ImportHistoryRow,
} from '@/types'

/** A refusal with a code, a sentence and a recovery — mapped by `lib/api/refusal.ts`. */
export class HistoryRefused extends Error {
  constructor(
    public code: string,
    message: string,
    public suggestion: string,
    public status: 400 | 404 | 409 = 400
  ) {
    super(message)
  }
}

// ---------------------------------------------------------------------------
// Checking a batch (pure — no database)
// ---------------------------------------------------------------------------

export interface RowProblem {
  /** Zero-based, as the mapper's array is. */
  index: number
  field: string
  reason: string
}

const ROW_KEYS = [
  'source',
  'source_ref',
  'company',
  'number',
  'client_name',
  'issue_date',
  'currency',
  'total',
  'status',
  'import_flag',
  'drive_path',
] as const

/** `numeric(14,2)`: twelve integer digits, at most two decimals, as a string. */
const MONEY = /^-?\d{1,12}(\.\d{1,2})?$/
/** The host `platform-storage` recognises as the blob store (`assets.ts`). */
const BLOB_HOST = 'blob.vercel-storage.com'

const SOURCES = new Set(HISTORY_SOURCES.map((t) => t.value))
const STATUSES = new Set(HISTORY_STATUSES.map((t) => t.value))

/** Not blank: at least one non-space character. The value itself is kept as given. */
function hasText(v: string): boolean {
  return /\S/.test(v)
}

/**
 * Check every row and return every problem, plus the rows typed.
 *
 * Checks what can be checked without the database: shapes, vocabularies,
 * lengths, dates, and a `(source, source_ref)` appearing twice IN THIS BATCH.
 * Whether the company exists and whether a row is already archived are
 * `importHistory`'s questions.
 */
export function checkImportRows(
  raw: unknown[],
  now: Date = new Date()
): { rows: ImportHistoryRow[]; problems: RowProblem[] } {
  const problems: RowProblem[] = []
  const today = todayInZurich(now)
  const seen = new Map<string, number>()

  raw.forEach((r, index) => {
    const bad = (field: string, reason: string) => problems.push({ index, field, reason })
    if (!r || typeof r !== 'object' || Array.isArray(r)) {
      bad('(row)', 'is not an object')
      return
    }
    const row = r as Record<string, unknown>

    for (const k of Object.keys(row)) {
      if (!(ROW_KEYS as readonly string[]).includes(k)) {
        bad(k, 'is not a field of an imported bill')
      }
    }

    const str = (k: string, max?: number): string | null => {
      const v = row[k]
      if (v === undefined || v === null) {
        bad(k, 'is required')
        return null
      }
      if (typeof v !== 'string') {
        bad(k, `is ${typeof v}; it must be a string`)
        return null
      }
      if (!hasText(v)) {
        bad(k, 'is blank')
        return null
      }
      if (max !== undefined && v.length > max) {
        bad(k, `is ${v.length} characters; the limit is ${max}`)
        return null
      }
      return v
    }

    const source = str('source')
    if (source !== null && !SOURCES.has(source)) {
      bad('source', `${JSON.stringify(source)} is not a source this app knows`)
    }
    const sourceRef = str('source_ref', HISTORY_LIMITS.source_ref_max)
    str('company')
    str('number', HISTORY_LIMITS.number_max)
    str('client_name', HISTORY_LIMITS.client_name_max)

    const issueDate = str('issue_date')
    if (issueDate !== null) {
      const d = new Date(`${issueDate}T00:00:00Z`)
      if (!/^\d{4}-\d{2}-\d{2}$/.test(issueDate) || Number.isNaN(d.getTime()) || d.toISOString().slice(0, 10) !== issueDate) {
        bad('issue_date', `${JSON.stringify(issueDate)} is not a calendar date as YYYY-MM-DD`)
      } else if (issueDate > today) {
        bad('issue_date', `${issueDate} is in the future; an archived bill was issued in the past`)
      }
    }

    const currency = str('currency')
    if (currency !== null && !/^[A-Z]{3}$/.test(currency)) {
      bad('currency', `${JSON.stringify(currency)} is not an ISO 4217 code in capitals`)
    }

    // Money is a STRING. A JSON number has already been through a float64 by
    // the time it reaches us, and "3240.1" and "3240.10000000000002" are the
    // same double — so a number is refused rather than formatted.
    const total = row.total
    if (typeof total === 'number') {
      bad('total', `is a JSON number; send it as a string, e.g. "${total.toFixed(2)}" — money is never a float here`)
    } else if (str('total') !== null && !MONEY.test(total as string)) {
      bad('total', `${JSON.stringify(total)} is not an amount with at most two decimals`)
    }

    const status = str('status')
    if (status !== null && !STATUSES.has(status)) {
      bad('status', `${JSON.stringify(status)} is not a history status`)
    }

    const flag = row.import_flag
    if (flag !== undefined && flag !== null) {
      if (typeof flag !== 'object' || Array.isArray(flag)) {
        bad('import_flag', 'must be {"fr": "…", "en": "…"} or null')
      } else {
        const f = flag as Record<string, unknown>
        for (const k of Object.keys(f)) {
          if (k !== 'fr' && k !== 'en') bad(`import_flag.${k}`, 'is not a language this app flags in (fr, en)')
        }
        for (const lang of ['fr', 'en'] as const) {
          const v = f[lang]
          if (typeof v !== 'string' || !hasText(v)) {
            // Both or neither. A flag in one language is hidden from every
            // reader of the other, and a hidden warning is not a warning.
            bad(`import_flag.${lang}`, 'is missing; a flag is given in both languages or not at all')
          } else if (v.length > HISTORY_LIMITS.import_flag_max) {
            bad(`import_flag.${lang}`, `is ${v.length} characters; the limit is ${HISTORY_LIMITS.import_flag_max}`)
          }
        }
      }
    }

    const drive = row.drive_path
    if (drive !== undefined && drive !== null) {
      if (typeof drive !== 'string' || !hasText(drive)) {
        bad('drive_path', 'must be a path or id on Google Drive, or null when the export had no PDF')
      } else if (drive.length > HISTORY_LIMITS.drive_path_max) {
        bad('drive_path', `is ${drive.length} characters; the limit is ${HISTORY_LIMITS.drive_path_max}`)
      } else if (drive.toLowerCase().includes(BLOB_HOST)) {
        bad('drive_path', 'points at this platform’s blob store; an archived PDF stays on Drive and this app stores a path, never a file')
      }
    }

    if (source !== null && sourceRef !== null) {
      const key = `${source} ${sourceRef}`
      const first = seen.get(key)
      if (first !== undefined) {
        bad('source_ref', `repeats row ${first}'s ${source} ${JSON.stringify(sourceRef)}; one source id is one bill`)
      } else {
        seen.set(key, index)
      }
    }
  })

  return { rows: raw as ImportHistoryRow[], problems }
}

/** One line per problem, capped, for a message a person reads on stderr. */
export function describeProblems(problems: RowProblem[], cap = 10): string {
  const shown = problems.slice(0, cap).map((p) => `row ${p.index}: ${p.field} ${p.reason}`)
  const more = problems.length > cap ? `; and ${problems.length - cap} more` : ''
  return shown.join('; ') + more
}

// ---------------------------------------------------------------------------
// Import
// ---------------------------------------------------------------------------

interface Existing {
  source: string
  source_ref: string
  seq: number
}

/** Rows of this batch already in the archive, by exact `(source, source_ref)`. */
async function alreadyImported(workspaceId: number, rows: ImportHistoryRow[]): Promise<Existing[]> {
  const refs = [...new Set(rows.map((r) => r.source_ref))]
  const found = await getDb()
    .select({ source: billingHistory.source, source_ref: billingHistory.source_ref, seq: billingHistory.seq })
    .from(billingHistory)
    .where(and(eq(billingHistory.workspace_id, workspaceId), inArray(billingHistory.source_ref, refs)))
  const wanted = new Set(rows.map((r) => `${r.source} ${r.source_ref}`))
  return found.filter((f) => wanted.has(`${f.source} ${f.source_ref}`))
}

function alreadyImportedRefusal(existing: Existing[], batchSize: number): HistoryRefused {
  const cap = 10
  const named = existing
    .slice(0, cap)
    .map((e) => `${e.source} ${JSON.stringify(e.source_ref)} is #${e.seq}`)
    .join('; ')
  const more = existing.length > cap ? `; and ${existing.length - cap} more` : ''
  return new HistoryRefused(
    'already_imported',
    `${existing.length} of ${batchSize} rows are already in the archive, so nothing was written: ${named}${more}`,
    'remove those rows and import the rest; if the source now says something different about one of them, that is for a person to look at — bk billing history show <#>',
    409
  )
}

/** A unique violation on `uq_history_source_ref`, anywhere on the cause chain. */
function isDuplicateSourceRef(err: unknown): boolean {
  for (let x = err as { code?: unknown; constraint?: unknown; cause?: unknown } | null | undefined; x; x = x.cause as typeof x) {
    if (x.code === '23505' && x.constraint === 'uq_history_source_ref') return true
  }
  return false
}

/**
 * Import a batch. All of it, or none of it.
 *
 * The order of refusals is the order a mapper can act on: the body's shape,
 * then every row problem at once, then unknown companies, then rows already in
 * the archive. Only then is anything allocated or written.
 */
export async function importHistory(ctx: WriteCtx, body: unknown, now: Date = new Date()): Promise<ImportHistoryResult> {
  const rawRows =
    body && typeof body === 'object' && !Array.isArray(body) ? (body as Record<string, unknown>).rows : undefined
  if (!Array.isArray(rawRows)) {
    throw new HistoryRefused(
      'invalid_body',
      'the body is {"rows": [ … ]}, one object per imported bill',
      'bk billing history import --file rows.json, where the file holds that object or the bare array'
    )
  }
  if (rawRows.length === 0) {
    throw new HistoryRefused('nothing_to_import', 'rows is empty', 'map at least one bill from the export')
  }
  if (rawRows.length > HISTORY_LIMITS.import_max_rows) {
    throw new HistoryRefused(
      'too_many_rows',
      `${rawRows.length} rows in one import; the limit is ${HISTORY_LIMITS.import_max_rows}`,
      'split the export into several imports — each is all-or-nothing, so a split never half-lands'
    )
  }

  const { rows, problems } = checkImportRows(rawRows, now)
  if (problems.length > 0) {
    throw new HistoryRefused(
      'invalid_history_rows',
      `${problems.length} problem${problems.length === 1 ? '' : 's'} in ${rawRows.length} rows, so nothing was written: ${describeProblems(problems)}`,
      'fix the rows named and import the whole file again; `bk meta --app-server billing` has the sources, statuses and limits'
    )
  }

  // Companies by slug, in this workspace only. A retired company is accepted:
  // it issued these bills when it was not retired, and a statement for a past
  // year has to be able to name it.
  const slugs = [...new Set(rows.map((r) => r.company))]
  const companies = await getDb()
    .select({ id: billingCompany.id, slug: billingCompany.slug })
    .from(billingCompany)
    .where(and(eq(billingCompany.workspace_id, ctx.workspaceId), inArray(billingCompany.slug, slugs)))
  const companyId = new Map(companies.map((c) => [c.slug, c.id]))
  const unknown = rows.map((r, i) => ({ r, i })).filter(({ r }) => !companyId.has(r.company))
  if (unknown.length > 0) {
    throw new HistoryRefused(
      'unknown_company',
      `${unknown.length} row${unknown.length === 1 ? '' : 's'} name a company this workspace does not have, so nothing was written: ` +
        describeProblems(unknown.map(({ r, i }) => ({ index: i, field: 'company', reason: JSON.stringify(r.company) }))),
      'bk billing company list — an issuing entity the mapper had to guess is still one of these, with the guess in import_flag'
    )
  }

  const existing = await alreadyImported(ctx.workspaceId, rows)
  if (existing.length > 0) throw alreadyImportedRefusal(existing, rows.length)

  let inserted: number[]
  try {
    inserted = await getDb().transaction(async (tx) => {
      const first = await allocateSeqBlock(tx, ctx.workspaceId, 'history', rows.length)
      await tx.insert(billingHistory).values(
        rows.map((r, i) => ({
          workspace_id: ctx.workspaceId,
          seq: first + i,
          source: r.source,
          source_ref: r.source_ref,
          company_id: companyId.get(r.company)!,
          number: r.number,
          client_name: r.client_name,
          issue_date: r.issue_date,
          currency: r.currency,
          total: r.total,
          status: r.status,
          import_flag_fr: r.import_flag?.fr ?? null,
          import_flag_en: r.import_flag?.en ?? null,
          drive_path: r.drive_path ?? null,
          imported_by: ctx.actorUserId,
          imported_via: ctx.via,
        }))
      )
      return rows.map((_, i) => first + i)
    })
  } catch (e) {
    // Lost a race to another import of the same row: its transaction committed
    // between our read and our insert. Answer as if we had seen it first.
    if (isDuplicateSourceRef(e)) {
      const now2 = await alreadyImported(ctx.workspaceId, rows)
      if (now2.length > 0) throw alreadyImportedRefusal(now2, rows.length)
    }
    throw e
  }

  const written = await listHistoryBySeq(ctx.workspaceId, inserted)
  return {
    imported: written.length,
    flagged: written.filter((w) => w.import_flag !== null).length,
    without_pdf: written.filter((w) => w.drive_path === null).length,
    rows: written,
  }
}

// ---------------------------------------------------------------------------
// Reads
// ---------------------------------------------------------------------------

const COLS = {
  seq: billingHistory.seq,
  source: billingHistory.source,
  source_ref: billingHistory.source_ref,
  company: billingCompany.slug,
  number: billingHistory.number,
  client_name: billingHistory.client_name,
  issue_date: billingHistory.issue_date,
  currency: billingHistory.currency,
  total: billingHistory.total,
  status: billingHistory.status,
  import_flag_fr: billingHistory.import_flag_fr,
  import_flag_en: billingHistory.import_flag_en,
  drive_path: billingHistory.drive_path,
  imported_at: billingHistory.imported_at,
  imported_by: billingHistory.imported_by,
  imported_by_email: users.email,
  imported_via: billingHistory.imported_via,
}

type Row = {
  seq: number
  source: string
  source_ref: string
  company: string
  number: string
  client_name: string
  issue_date: string
  currency: string
  total: string
  status: string
  import_flag_fr: string | null
  import_flag_en: string | null
  drive_path: string | null
  imported_at: Date
  imported_by: number | null
  imported_by_email: string | null
  imported_via: string
}

function shape(r: Row): HistoryEntry {
  return {
    seq: r.seq,
    source: r.source as HistorySource,
    source_ref: r.source_ref,
    company: r.company,
    number: r.number,
    client_name: r.client_name,
    issue_date: r.issue_date,
    currency: r.currency,
    total: r.total,
    status: r.status as HistoryStatus,
    // The CHECK makes these both-or-neither; read defensively anyway, so a row
    // that somehow had one would still SHOW its flag rather than drop it.
    import_flag:
      r.import_flag_fr !== null || r.import_flag_en !== null
        ? { fr: r.import_flag_fr ?? r.import_flag_en ?? '', en: r.import_flag_en ?? r.import_flag_fr ?? '' }
        : null,
    drive_path: r.drive_path,
    imported_at: r.imported_at.toISOString(),
    imported_by: { user_id: r.imported_by, email: r.imported_by_email, via: r.imported_via as ActorVia },
  }
}

function base() {
  return (
    getDb()
      .select(COLS)
      .from(billingHistory)
      .innerJoin(billingCompany, eq(billingCompany.id, billingHistory.company_id))
      // LEFT: `imported_by` is ON DELETE SET NULL, and a row whose importer is
      // gone must still be listed. An inner join would make it vanish.
      .leftJoin(users, eq(users.id, billingHistory.imported_by))
  )
}

async function listHistoryBySeq(workspaceId: number, seqs: number[]): Promise<HistoryEntry[]> {
  const rows = (await base()
    .where(and(eq(billingHistory.workspace_id, workspaceId), inArray(billingHistory.seq, seqs)))
    .orderBy(billingHistory.seq)) as Row[]
  return rows.map(shape)
}

export interface ListHistoryOptions {
  source?: HistorySource
  currency?: string
  /** A calendar year: bills issued in it. */
  year?: number
  /** Only rows carrying an import flag. */
  flagged?: boolean
  company?: string
  limit?: number
  /** The `seq` of the last row of the previous page. */
  cursor?: number
}

/**
 * Newest bill first (`issue_date`, then `#number`), which is the order the
 * screen groups by year in.
 *
 * ── THE CURSOR IS A #NUMBER, AND THE ORDER IS BY DATE ─────────────────────
 * The next page is "everything after the row with this #number, in list order",
 * and that row's date is read back to continue from. Stable because the rows
 * are read-only: the row the cursor names cannot have moved since the last
 * page. An unknown cursor is a 400 rather than page one again, which a paginator
 * would loop on forever.
 */
export async function listHistory(
  workspaceId: number,
  opts: ListHistoryOptions = {}
): Promise<{ data: HistoryEntry[]; next_cursor: number | null }> {
  const limit = Math.min(Math.max(opts.limit ?? LIST_LIMIT_DEFAULT, 1), LIST_LIMIT_MAX)
  const where = [eq(billingHistory.workspace_id, workspaceId)]
  if (opts.source) where.push(eq(billingHistory.source, opts.source))
  if (opts.currency) where.push(eq(billingHistory.currency, opts.currency))
  if (opts.company) where.push(eq(billingCompany.slug, opts.company))
  if (opts.year !== undefined) {
    where.push(sql`${billingHistory.issue_date} >= ${`${opts.year}-01-01`}::date`)
    where.push(sql`${billingHistory.issue_date} < ${`${opts.year + 1}-01-01`}::date`)
  }
  if (opts.flagged) where.push(sql`${billingHistory.import_flag_en} IS NOT NULL`)
  if (opts.cursor !== undefined) {
    const at = await getDb()
      .select({ issue_date: billingHistory.issue_date, seq: billingHistory.seq })
      .from(billingHistory)
      .where(and(eq(billingHistory.workspace_id, workspaceId), eq(billingHistory.seq, opts.cursor)))
      .limit(1)
    if (at.length === 0) {
      throw new HistoryRefused(
        'invalid_cursor',
        `no imported bill #${opts.cursor} in this workspace to continue after`,
        'pass the next_cursor the previous page returned, or drop it to start from the newest'
      )
    }
    where.push(sql`(${billingHistory.issue_date}, ${billingHistory.seq}) < (${at[0].issue_date}::date, ${at[0].seq})`)
  }

  const rows = (await base()
    .where(and(...where))
    .orderBy(desc(billingHistory.issue_date), desc(billingHistory.seq))
    .limit(limit + 1)) as Row[]
  const page = rows.slice(0, limit)
  return {
    data: page.map(shape),
    next_cursor: rows.length > limit ? (page[page.length - 1]?.seq ?? null) : null,
  }
}

/** One imported bill by its #number, or null. */
export async function getHistory(workspaceId: number, seq: number): Promise<HistoryEntry | null> {
  const rows = (await base()
    .where(and(eq(billingHistory.workspace_id, workspaceId), eq(billingHistory.seq, seq)))
    .limit(1)) as Row[]
  return rows[0] ? shape(rows[0]) : null
}

File: cli/internal/commands/billing/history.go
// `bk billing history …` — the imported archive (phase 5).
package billing

import (
	"encoding/json"
	"fmt"
	"io"
	"strconv"
	"strings"

	"github.com/blackcode-switzerland/bc-issues/cli/internal/client"
	"github.com/blackcode-switzerland/bc-issues/cli/internal/cmdutil"
	"github.com/blackcode-switzerland/bc-issues/cli/internal/output"
	"github.com/spf13/cobra"
)

func newHistoryCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "history",
		Short: "Bills issued before this app existed — read-only",
		Long: `The imported archive: bills from the systems this app replaces.

AN ARCHIVE, NOT MORE INVOICES. Each row keeps the number it was issued with, in
its own namespace — it never enters an issuing company's sequence, and its
number is not unique (two systems, two companies), so rows are addressed by
their #number. A row is written once and never edited or deleted; the database
refuses both.

YOU MAP, THE APP STORES. There is no importer and no mapping engine here. Read
the export, map each bill onto the row shape, and import the rows. Whatever you
could not resolve — a possible duplicate across a migration window, an issuing
company inferred from a bank account, a missing VAT amount — goes in the row's
import flag, in words, in both languages. It is shown on every surface and
never cleared.

THE ARCHIVED PDF STAYS ON DRIVE. A row stores its path or id there, or nothing
when the export had no PDF, and says so.

Sources, statuses and the per-import row limit: "bk meta --app-server billing".`,
	}
	cmd.AddCommand(newHistoryListCmd(), newHistoryShowCmd(), newHistoryImportCmd())
	return cmd
}

func newHistoryListCmd() *cobra.Command {
	var o client.ListBillingHistoryOptions
	cmd := &cobra.Command{
		Use:         "list",
		Annotations: map[string]string{"routes": "GET /api/workspaces/{ws}/history"},
		Short:       "The archive, newest bill first",
		Long: `Imported bills, newest issue date first.

--flagged lists only the rows carrying an import flag — the ones somebody still
has to look at. --year, --source, --currency and --company narrow it. A filter
the server does not understand is refused, never ignored.

The FLAG column shows the English text, shortened; "bk billing history show"
prints it in full in both languages.`,
		RunE: func(cmd *cobra.Command, args []string) error {
			format, err := output.Resolve(cmd)
			if err != nil {
				return err
			}
			c, ws, err := clientAndWorkspace()
			if err != nil {
				return err
			}
			page, err := c.ListBillingHistory(ws, o)
			if err != nil {
				return err
			}
			return output.Render(format, page, func(w io.Writer) error {
				if len(page.Data) == 0 {
					fmt.Fprintln(w, "Nothing in the archive matches.")
					nextStep(w, "bk billing history import --file rows.json")
					return nil
				}
				tw := output.Tabwriter(w)
				fmt.Fprintln(tw, "#\tISSUED\tSOURCE\tSOURCE REF\tNUMBER\tCOMPANY\tCLIENT\tTOTAL\tSTATUS\tPDF\tFLAG")
				for _, h := range page.Data {
					fmt.Fprintf(tw, "%d\t%s\t%s\t%s\t%s\t%s\t%s\t%s %s\t%s\t%s\t%s\n",
						h.Number, h.IssueDate, h.Source, quoteIfPadded(h.SourceRef), h.Ref, h.Company,
						cmdutil.Truncate(h.ClientName, 22), h.Currency, h.Total, h.Status,
						pdfCell(h.DrivePath), flagCell(h.Flag, 36))
				}
				if err := tw.Flush(); err != nil {
					return err
				}
				if page.NextCursor != nil {
					fmt.Fprintf(w, "\nmore: --cursor %d\n", *page.NextCursor)
				}
				return nil
			})
		},
	}
	f := cmd.Flags()
	f.StringVar(&o.Source, "source", "", "Only this source (bk meta for the values)")
	f.StringVar(&o.Currency, "currency", "", "Only this currency")
	f.StringVar(&o.Company, "company", "", "Only this company's bills")
	f.IntVar(&o.Year, "year", 0, "Only bills issued in this calendar year")
	f.BoolVar(&o.Flagged, "flagged", false, "Only rows carrying an import flag")
	f.IntVar(&o.Limit, "limit", 0, "How many (bk meta for the ceiling)")
	f.IntVar(&o.Cursor, "cursor", 0, "Continue from the cursor the last page returned")
	return cmd
}

func newHistoryShowCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:         "show <#>",
		Annotations: map[string]string{"routes": "GET /api/workspaces/{ws}/history/{seq}"},
		Args:        cobra.ExactArgs(1),
		Short:       "One imported bill, its flag in both languages, where its PDF is",
		Long: `One imported bill by its #number — the number "history list" shows in its
first column. The historical number is not unique, so it is not accepted here.`,
		RunE: func(cmd *cobra.Command, args []string) error {
			format, err := output.Resolve(cmd)
			if err != nil {
				return err
			}
			seq, err := strconv.Atoi(strings.TrimPrefix(args[0], "#"))
			if err != nil || seq < 1 {
				return fmt.Errorf("%q is not a #number — run \"bk billing history list\" for them", args[0])
			}
			c, ws, err := clientAndWorkspace()
			if err != nil {
				return err
			}
			h, err := c.GetBillingHistory(ws, seq)
			if err != nil {
				return err
			}
			return output.Render(format, h, func(w io.Writer) error {
				fmt.Fprintf(w, "#%d  %s  (imported from %s)\n", h.Number, h.Ref, h.Source)
				tw := output.Tabwriter(w)
				fmt.Fprintf(tw, "Source ref:\t%s\n", quoteIfPadded(h.SourceRef))
				fmt.Fprintf(tw, "Company:\t%s\n", h.Company)
				fmt.Fprintf(tw, "Client:\t%s\n", h.ClientName)
				fmt.Fprintf(tw, "Issued:\t%s\n", h.IssueDate)
				fmt.Fprintf(tw, "Total:\t%s %s\n", h.Currency, h.Total)
				fmt.Fprintf(tw, "Status:\t%s\n", h.Status)
				if h.DrivePath != nil {
					fmt.Fprintf(tw, "PDF:\t%s\n", *h.DrivePath)
				} else {
					fmt.Fprintf(tw, "PDF:\tno PDF in the export\n")
				}
				who := h.ImportedBy.Email
				if who == "" {
					who = "(account removed)"
				}
				fmt.Fprintf(tw, "Imported:\t%s by %s via %s\n", h.ImportedAt[:19], who, h.ImportedBy.Via)
				if err := tw.Flush(); err != nil {
					return err
				}
				if h.Flag != nil {
					fmt.Fprintf(w, "\nFLAG (en): %s\nFLAG (fr): %s\n", h.Flag.En, h.Flag.Fr)
				} else {
					fmt.Fprintln(w, "\nNo import flag: the row mapped cleanly.")
				}
				return nil
			})
		},
	}
	return cmd
}

func newHistoryImportCmd() *cobra.Command {
	var file string
	cmd := &cobra.Command{
		Use:         "import --file rows.json",
		Annotations: map[string]string{"routes": "POST /api/workspaces/{ws}/history"},
		Short:       "Import mapped rows — all of them, or none",
		Long: `Import bills you have mapped from an export.

--file is a JSON array of rows, or an object {"rows": [...]}. "--file -" reads
it from stdin, which is how to pipe a mapped batch without quoting JSON through
a shell.

ONE ROW:

  {
    "source": "zoho",
    "source_ref": "ZB-000178",
    "company": "blackcode",
    "number": "2023-011",
    "client_name": "Transports Junod SA",
    "issue_date": "2023-12-01",
    "currency": "CHF",
    "total": "21500.00",
    "status": "paid",
    "import_flag": {"fr": "…", "en": "…"},
    "drive_path": "Archive/Zoho/2023/ZB-000178.pdf"
  }

"total" is a STRING. A JSON number is refused: it has been through a float by
the time it arrives. "source_ref" is stored exactly as given — not trimmed, not
reformatted — because it is the only way back to the source. "import_flag" is
both languages or omitted; "drive_path" is omitted or null when the export had
no PDF. A key the row shape does not have is refused, so a typo cannot drop a
field on the way in.

ALL OR NOTHING. Every row is checked and every problem is reported at once; if
any row is refused, nothing is written. A row already in the archive — same
source, same source ref — refuses the import and names the #number it already
has. Importing the same file twice therefore changes nothing and says which
rows were there. It is a refusal rather than a skip on purpose: if the export
now says something different about a bill, a person should look.

The output names every row written, with its new #number.`,
		RunE: func(cmd *cobra.Command, args []string) error {
			format, err := output.Resolve(cmd)
			if err != nil {
				return err
			}
			if file == "" {
				return fmt.Errorf("--file is required: a JSON file of mapped rows, or \"-\" for stdin")
			}
			var raw string
			if file == "-" {
				raw, err = cmdutil.ReadBody("-", "")
			} else {
				raw, err = cmdutil.ReadBody("", file)
			}
			if err != nil {
				return err
			}
			rows, err := rowsFromFile([]byte(raw))
			if err != nil {
				return err
			}
			c, ws, err := clientAndWorkspace()
			if err != nil {
				return err
			}
			res, err := c.ImportBillingHistory(ws, rows)
			if err != nil {
				return err
			}
			return output.Render(format, res, func(w io.Writer) error {
				fmt.Fprintf(w, "Imported %d bill%s as %s: %d flagged, %d without a PDF.\n",
					res.Imported, plural(res.Imported), numberRange(res.Rows), res.Flagged, res.WithoutPdf)
				tw := output.Tabwriter(w)
				fmt.Fprintln(tw, "#\tSOURCE\tSOURCE REF\tNUMBER\tISSUED\tTOTAL\tFLAG")
				for _, h := range res.Rows {
					fmt.Fprintf(tw, "%d\t%s\t%s\t%s\t%s\t%s %s\t%s\n",
						h.Number, h.Source, quoteIfPadded(h.SourceRef), h.Ref, h.IssueDate,
						h.Currency, h.Total, flagCell(h.Flag, 36))
				}
				if err := tw.Flush(); err != nil {
					return err
				}
				if res.Flagged > 0 {
					nextStep(w, "bk billing history list --flagged")
				} else {
					nextStep(w, "bk billing history list")
				}
				return nil
			})
		},
	}
	cmd.Flags().StringVar(&file, "file", "", `JSON array of rows, or {"rows": [...]}; "-" reads stdin`)
	return cmd
}

// rowsFromFile accepts a bare array or {"rows": [...]} and returns the array as
// RAW JSON — never decoded into Go structs, so a key the mapper misspelled
// reaches the server and is refused there, rather than being dropped here.
func rowsFromFile(b []byte) (json.RawMessage, error) {
	trimmed := strings.TrimSpace(string(b))
	if strings.HasPrefix(trimmed, "[") {
		var probe []json.RawMessage
		if err := json.Unmarshal([]byte(trimmed), &probe); err != nil {
			return nil, fmt.Errorf("the file is not a JSON array: %v", err)
		}
		return json.RawMessage(trimmed), nil
	}
	var wrapped struct {
		Rows json.RawMessage `json:"rows"`
	}
	if err := json.Unmarshal([]byte(trimmed), &wrapped); err != nil {
		return nil, fmt.Errorf("the file is neither a JSON array of rows nor {\"rows\": [...]}: %v", err)
	}
	if len(wrapped.Rows) == 0 || !strings.HasPrefix(strings.TrimSpace(string(wrapped.Rows)), "[") {
		return nil, fmt.Errorf("the file is an object without a \"rows\" array")
	}
	return wrapped.Rows, nil
}

// quoteIfPadded shows a source ref in quotes when it has leading or trailing
// whitespace, which a table would otherwise make invisible. The value itself is
// never changed.
func quoteIfPadded(s string) string {
	if s != strings.TrimSpace(s) {
		return strconv.Quote(s)
	}
	return s
}

func pdfCell(drivePath *string) string {
	if drivePath == nil {
		return "none"
	}
	return "drive"
}

func flagCell(f *client.BillingImportFlag, n int) string {
	if f == nil {
		return "—"
	}
	return "⚑ " + cmdutil.Truncate(f.En, n)
}

// numberRange is "#15", or "#1–#14" for a block. An import's numbers are
// contiguous — one allocation for the whole batch — so a range names them all.
func numberRange(rows []client.BillingHistoryEntry) string {
	if len(rows) == 0 {
		return "nothing"
	}
	first, last := rows[0].Number, rows[len(rows)-1].Number
	if first == last {
		return fmt.Sprintf("#%d", first)
	}
	return fmt.Sprintf("#%d–#%d", first, last)
}

func plural(n int) string {
	if n == 1 {
		return ""
	}
	return "s"
}
